libazureinit/src/provision/user.rs

```diff
@@ -123,10 +127,39 @@ fn useradd(user: &User) -> Result<(), Error> {
     Ok(())
 }
 
+fn add_user_for_passwordless_sudo(
+    username: String,
+    path: &str,
+) -> Result<(), Error> {
+    // Create a file under /etc/sudoers.d with azure-init-user
+    let sudoers_path = path;
+    let mut sudoers_file = std::fs::OpenOptions::new()
+        .append(true)
+        .create(true)
+        .open(sudoers_path)?;
+    write!(
+        sudoers_file,
+        "{} ALL=(ALL) NOPASSWD: ALL \n",
+        username.to_string()
+    )?;
+    sudoers_file.flush()?;
+    // Set the permission
+    sudoers_file.set_permissions(Permissions::from_mode(0o600))?;
```

Comment:
My concern with this approach is that there's a window between creating the file and setting the permissions that it is potentially world-writable. Additionally, the file has been opened to append, meaning we won't clobber any write someone slips in before the permissions get set appropriately.

Now, _presumably_ this runs early in boot on a trusted image, but we don't know how folks may use the library and violate that expectation. I wasn't aware of this or I would have suggested it earlier, but there's a [Unix extension in the standard library](https://doc.rust-lang.org/std/os/unix/fs/trait.OpenOptionsExt.html), so I recommend doing this instead:
```
use std::os::unix::fs::OpenOptionsExt;

std::fs::OpenOptions::new()
    .mode(0o600)
    .truncate(true)
    .create(true)
    .open(sudoers_path)?;
```

I think we would prefer truncating the file if it already exists rather than accepting any previous file since it's a drop-in file specific to this program. You could imagine someone building an image from one that got provisioned with this tool and failing to clean up this configuration. We wouldn't want to leave that original provisioning user with sudo permissions.

Reply:
Done